python/composio/tools/env/host/shell.py

```diff
@@ -28,13 +33,42 @@ def __init__(self) -> None:
             text=True,
             bufsize=1,
         )
-        self.logger.debug(f"Initial data from session: {self.id} - {self._read()}")
+        self.logger.debug(
+            f"Initial data from session: {self.id} - {self._read(wait=False)}"
+        )
+
+    def _wait_for_cmd(self, cmd: str) -> None:
+        """Waif for command to exit."""
+        while True:
+            output = subprocess.run(  # pylint: disable=subprocess-run-check
+                ["ps", "-e"],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+            ).stdout.decode()
+            if all(_cmd.lstrip().rstrip() not in output for _cmd in cmd.split("&&")):
+                return
+            time.sleep(1)
+
+    def _get_exit_code(self) -> int:
+        """Get exit code of the last process."""
+        self._write("echo $#")
+        return int(self._read(wait=False).get("stdout").strip())  # type: ignore
 
-    def _read(self, timeout: float = 120.0) -> t.Dict:
+    def _read(
+        self,
+        cmd: t.Optional[str] = None,
+        wait: bool = True,
+        timeout: float = 120.0,
+    ) -> t.Dict:
         """Read data from a subprocess with a timeout."""
         stderr = t.cast(t.IO[str], self._process.stderr).fileno()
         stdout = t.cast(t.IO[str], self._process.stdout).fileno()
         buffer = {stderr: b"", stdout: b""}
+        if wait and cmd is None:
+            raise ValueError("`cmd` cannot be `None` when `wait` is set to `True`")
+
+        if wait:
```

Comment:
what is the point of having "wait" and "timout" both ?

Reply:
Fixed - https://github.com/ComposioHQ/composio/pull/268/commits/89336b1b091e40b78a9d4e862ce9c35b9ab1fdd6#diff-4c160bb7df37eb11849385599ab5110fcf0628e2ed1c7a0950e0365b90fa5bfaR70-R72